core/src/core/transaction.rs

```diff
@@ -379,35 +394,108 @@ impl Transaction {
+/// But also information required to verify coinbase maturity through
+/// the lock_height hashed in the switch_commit_hash.
+#[derive(Debug, Clone, Copy)]
+pub struct Input{
+	commit: Commitment,
+	// We need to provide the switch_commit (switch_commit_hash preimage)
+	// to verify coinbase maturity.
+	switch_commit: Commitment,
+	// We need to provide lock_height to verify coinbase maturity.
+	lock_height: u64,
+}
 
 hashable_ord!(Input);
 
 /// Implementation of Writeable for a transaction Input, defines how to write
 /// an Input as binary.
 impl Writeable for Input {
 	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ser::Error> {
-		writer.write_fixed_bytes(&self.0)
+		writer.write_fixed_bytes(&self.commit)?;
+		writer.write_fixed_bytes(&self.switch_commit)?;
+		writer.write_u64(self.lock_height)?;
+		Ok(())
 	}
 }
 
 /// Implementation of Readable for a transaction Input, defines how to read
 /// an Input from a binary stream.
 impl Readable for Input {
 	fn read(reader: &mut Reader) -> Result<Input, ser::Error> {
-		Ok(Input(Commitment::read(reader)?))
+		Ok(Input {
+			commit: Commitment::read(reader)?,
+			switch_commit: Commitment::read(reader)?,
+			lock_height: reader.read_u64()?,
+		})
 	}
 }
 
 /// The input for a transaction, which spends a pre-existing output. The input
 /// commitment is a reproduction of the commitment of the output it's spending.
 impl Input {
+	/// Build a new Input from a commit, switch_commit and lock_height
+	pub fn new(
+		commit: Commitment,
+		switch_commit: Commitment,
+		lock_height: u64,
+	) -> Input {
+		debug!(
+			LOGGER,
+			"building a new input: {:?}, {:?}, {:?}",
+			commit,
+			switch_commit,
+			lock_height,
+		);
+		Input {
+			commit,
+			switch_commit,
+			lock_height,
+		}
+	}
+
 	/// Extracts the referenced commitment from a transaction output
 	pub fn commitment(&self) -> Commitment {
-		self.0
+		self.commit
+	}
+
+	/// The lock_height of this input (only relevant for an input spending a coinbase output).
+	pub fn lock_height(&self) -> u64 {
+		self.lock_height
+	}
+
+	/// Given the output being spent by this input along with the current chain height
+	/// we can verify the output has sufficiently matured and is spendable.
+	/// We do this by reconstructing the switch_commit_hash from the switch_commit.
+	/// The lock_height is passed in as the secret key to the hash so we can verify
+	/// we are not being lied to about the lock_height.
+	///
+	/// Note: we are only enforcing this for coinbase outputs currently.
+	///
+	pub fn verify_lock_height(
+		&self,
+		output_switch_commit_hash: &SwitchCommitHash,
+		height: u64,
+	) -> Result<(), Error> {
+		debug!(LOGGER, "verify_lock_height: {}, {}", self.lock_height, height);
+		
+		let switch_commit_hash = SwitchCommitHash::from_switch_commit(
+			self.switch_commit,
+			SwitchCommitHashKey::from_lock_height(self.lock_height),
+		);
+
+		if switch_commit_hash != *output_switch_commit_hash {
```

Comment:
Similarly, I thought we would only check the switch commitment pre-image match when spending coinbase outputs for now? And then we can soft fork it for all inputs when Quantumageddon is near.

Reply:
There was some reason why I thought this wouldn't work - can't remember what it is (and I'm probably mistaken). Trying to remember what it was...

Reply:
Don't we need to check if it's spending a coinbase output anyway? Otherwise a miner could include a lock time of 1 in the coinbase output, which we can't know, and then spend it after 2 blocks.

Reply:
Oof, yes of course - we were originally (in this PR) checking that we were not being lied to about lock_height by finding the output being spent and verifying the height of that block was consistent with what the lock_height said.